Remove explicit time calculations from Song and TimeLineWidget

michaelgregorius opened the pull request:
Removes some explicit time calculations from `Song` and `TimeLineWidget` as part of #3284. See [this comment](https://github.com/LMMS/lmms/issues/3284#issuecomment-315542540).

PhysSong commented:
Good. Could you also fix two more places where `* 60 * 1000 / 48` appears? They are in Song.cpp and you can use `MidiTime::ticksToMilliseconds()`.

PhysSong commented:
Coding conventions looks inconsistent a little bit but isn't wrong.

michaelgregorius commented:
@PhysSong Good catch! I had also checked the file for other occurrences (searching for "48" but somehow it did not occur to me that I can replace two more calls by using on of the methods that I had just added. 😄

I have removed two more calls and squashed the two commits into one.

PhysSong commented:
@michaelgregorius You don't need to squash commits manually. They can be squashed when PR is merged.

zonkmachine commented:
@PhysSong The main reason for squashing the commits yourself, the way I see it, is to have control of the final commit message. If someone else does it for you, which is a risk at places like this, they may just press merge and all the commit messages will be automatically stacked one after the other.

PhysSong commented:
@michaelgregorius I'm working on #2112. If some problems occur due to this change, I'll fix it. I suggest merging.

PhysSong commented:
I'm going to fix #2112. When I finish my work, I'll make a pull request against stable-1.2 and my PR needs this fix. Is it okay to rebase to stable-1.2 and change the base branch?

tresf commented:
> I'll make a pull request against stable-1.2 and my PR needs this fix. Is it okay to rebase to stable-1.2 and change the base branch?

If this bug is not new to 1.2, it shouldn't be targeted at 1.2.  We are trying to freeze the codebase.

michaelgregorius commented:
Are you sure that the fix for #2112 should be targeted for 1.2? It's not in the milestones for 1.2 and at a certain point we should really get finished with 1.2. By the way, even if this fix goes into master first we can always cherry-pick it for 1.2.

@tresf What do you think?

michaelgregorius commented:
@tresf Seems like we posted at the same time. :smile:

PhysSong commented:
Though it isn't a new bug and changes the codebase, it doesn't make something unstable or change the codebase a lot. Moreover, it is a bug fix.

tresf commented:
I still don't see a compelling reason to merge this into stable.  This code is called all over the place, I'd rather not change it at the RC4 state so close to release.

PhysSong commented:
Okay. Go to master.

PhysSong commented on include/Song.h:
```diff
@@ -101,14 +101,22 @@ class EXPORT Song : public TrackContainer
 	{
 		return m_elapsedMilliSeconds;
 	}
-	inline void setMilliSeconds( float ellapsedMilliSeconds )
```
I prefer leaving `setMilliSeconds()` to removing, but it's not bad. We can add this again when needed.

michaelgregorius replied:
@PhysSong Yes, we can always add it back if needed. However, for now it look like the milliseconds are rather an internal representation of the `Song` class and that other client "think" in different times, e.g. `MidiTime`.

karmux commented:
Time display is not changing during playing anymore with this change. Time only changes when you click somewhere on timeline but not during playing.

zonkmachine commented:
I see this too. I've compiled both with and without debug flags.

PhysSong commented:
>Time display is not changing during playing anymore with this change.

I haven't tested yet, but I will fix it if I can reproduce.

zonkmachine commented:
> I haven't tested yet, but I will fix it if I can reproduce.

I can't answer for @michaelgregorius but why not leave him some time to respond? I'm pretty sure he knows what's up with this one already.

PhysSong commented:
@zonkmachine I think @michaelgregorius can fix it. However, I haven't seen this after my work for #2112.
If @michaelgregorius can fix it soon, I'll let him do.

PhysSong commented on src/core/Song.cpp:
```diff
@@ -393,9 +388,7 @@ void Song::processNextBuffer()
-		m_elapsedMilliSeconds += 
-			( ( framesToPlay / framesPerTick ) * 60 * 1000 / 48 ) 
-				/ getTempo();
+		m_elapsedMilliSeconds += MidiTime::ticksToMilliseconds( framesToPlay / framesPerTick, getTempo());
```
I think this line caused bug. Possible fix is: change
`double MidiTime::ticksToMilliseconds(tick_t ticks, bpm_t beatsPerMinute)` to
`double MidiTime::ticksToMilliseconds(double ticks, bpm_t beatsPerMinute)`.

PhysSong replied:
Or adding `MidiTime::framesToMillisecond(frame_t ticks, const float framesPerTick, bpm_t beatsPerMiute)`?

michaelgregorius commented:
Problem is fixed with commit bcdb5ecb5af503839070a77edb23e1731c74856e. I have decided to add a second version of `MidiTime::ticksToMilliseconds` which takes a double as the argument for the ticks.

Thanks for the hint, @PhysSong!

zonkmachine commented:
Tested, fixed!

The somewhat spooky thing is that I actually started a review and the action around the timeline was the only thing I got around to test and still I didn't spot it... :/

michaelgregorius commented:
@zonkmachine Thanks for testing! Seems to have been quite some elusive bug then. I think I only tested whether the play head moves or not.
